src/backend/distributed/commands/multi_copy.c

```diff
@@ -369,39 +383,96 @@ CopyToExistingShards(CopyStmt *copyStatement, char *completionTag)
-		copiedDistributedRelationTuple = (Form_pg_class) palloc(CLASS_TUPLE_SIZE);
-
-		/*
-		 * There is no need to deep copy everything. We will just deep copy of the fields
-		 * we will change.
-		 */
-		memcpy(copiedDistributedRelation, distributedRelation, sizeof(RelationData));
-		memcpy(copiedDistributedRelationTuple, distributedRelation->rd_rel,
-			   CLASS_TUPLE_SIZE);
-
 		copiedDistributedRelationTuple->relkind = RELKIND_RELATION;
-		copiedDistributedRelation->rd_rel = copiedDistributedRelationTuple;
 	}
-	else
+
+	/*
+	 * We make an optimisation to skip JSON parsing for JSONB columns, because many
+	 * Citus users have large objects in this column and parsing it on the coordinator
+	 * causes significant CPU overhead. We do this by forcing BeginCopyFrom and
+	 * NextCopyFrom to parse the column as text and then encoding it as JSON again
+	 * by using citus_text_send_as_jsonb as the binary output function.
+	 *
+	 * The main downside of enabling this optimisation is that it defers validation
+	 * unti the object is parsed by the worker, which is unable to give an accurate
+	 * line number.
+	 */
+	if (SkipJsonbValidationInCopy && !isInputFormatBinary)
 	{
-		/*
-		 * If we are not dealing with partitioned table, copiedDistributedRelation is same
-		 * as distributedRelation.
-		 */
-		copiedDistributedRelation = distributedRelation;
+		CopyOutState copyOutState = copyDest->copyOutState;
+		List *jsonbColumnIndexList = NIL;
+		ListCell *jsonbColumnIndexCell = NULL;
+
+		/* get the column indices for all JSONB columns that appear in the input */
+		jsonbColumnIndexList = FindJsonbInputColumns(copiedDistributedRelation->rd_att,
+													 copyStatement->attlist);
+
+		foreach(jsonbColumnIndexCell, jsonbColumnIndexList)
+		{
+			int columnIndex = lfirst_int(jsonbColumnIndexCell);
+			Form_pg_attribute currentColumn =
+				TupleDescAttr(copiedDistributedRelation->rd_att, columnIndex);
+			FmgrInfo *currentOutputFunction =
+				&copyDest->columnOutputFunctions[columnIndex];
+
+			if (columnIndex == partitionColumnIndex)
+			{
+				/*
+				 * In the curious case of using a JSONB column as partition column,
+				 * we leave it as is because we want to make sure the hashing works
+				 * correctly.
+				 */
+				continue;
+			}
+
+			ereport(DEBUG1, (errmsg("parsing JSONB column %s as text",
+									NameStr(currentColumn->attname))));
+
+			/* parse the column as text instead of JSONB */
+			currentColumn->atttypid = TEXTOID;
+
+			if (copyOutState->binary)
+			{
+				Oid textSendAsJsonbFunctionId = CitusTextSendAsJsonbFunctionId();
+
+				/*
+				 * If we're using binary encoding between coordinator and workers
+				 * then we should honour the format expected by jsonb_recv, which
+				 * is a version number followed by text. We thefere use an output
+				 * function which sends the text as if it were jsonb, namely by
+				 * prepending a version number.
+				 */
+				fmgr_info(textSendAsJsonbFunctionId, currentOutputFunction);
```

Comment:
I think using `&copyDest->columnOutputFunctions[columnIndex]` directly instead of a variable might help the readability a bit. When I saw this I first expected a use of `currentOutputFunction` after this line, but after a bit of thinking I realized that we are actually modifying the array inside the `copyDest`. If we didn't use a variable, I needed less thinking.